meeteval/io/seglst.py

```diff
@@ -0,0 +1,388 @@
+        return key
+    elif isinstance(key, (str, int)):
+        return operator.itemgetter(key)
+    else:
+        raise TypeError(f'Invalid type for key: {type(key)}')
+
+
+def groupby(
+        iterable,
+        key=None,
+        default_key=None,
+):
+    """
+    A non-lazy variant of `itertools.groupby` with advanced features.
+
+    Copied from `paderbox.utils.iterable.groupby`.
+
+    Args:
+        iterable: Iterable to group
+        key: Determines by what to group. Can be:
+            - `None`: Use the iterables elements as keys directly
+            - `callable`: Gets called with every element and returns the group
+                key
+            - `str`, or `int`: Use `__getitem__` on elements in `iterable`
+                to obtain the key
+            - `Iterable`: Provides the keys. Has to have the same length as
+                `iterable`.
+
+    Examples:
+        >>> groupby('ab'*3)
+        {'a': ['a', 'a', 'a'], 'b': ['b', 'b', 'b']}
+        >>> groupby(range(10), lambda x: x%2)
+        {0: [0, 2, 4, 6, 8], 1: [1, 3, 5, 7, 9]}
+        >>> groupby(({'a': x%2, 'b': x} for x in range(3)), 'a')
+        {0: [{'a': 0, 'b': 0}, {'a': 0, 'b': 2}], 1: [{'a': 1, 'b': 1}]}
+        >>> groupby(['abc', 'bd', 'abd', 'cdef', 'c'], 0)
+        {'a': ['abc', 'abd'], 'b': ['bd'], 'c': ['cdef', 'c']}
+        >>> groupby('abc', {})
+        Traceback (most recent call last):
+            ...
+        TypeError: Invalid type for key: <class 'dict'>
+    """
+    import collections
+    import itertools
+
+    groups = collections.defaultdict(list)
+    try:
+        for key, group in itertools.groupby(iterable, _get_key(key)):
+            groups[key].extend(group)
+    except KeyError:
+        if default_key is None:
+            raise
+        else:
+            assert len(groups) == 0, (groups, iterable, key)
+            return iterable
+    return dict(groups)
+
+
+def seglst_map(*, required_keys=(), py_convert=NestedStructure):
+    """
+    Decorator to for a function that takes a (single) `SegLST` object as input and returns a (single) `SegLST` object
+    as output. Automatically converts the input to `SegLST` and converts the returned value back to its original type.
+
+    >>> @seglst_map(required_keys=('speaker',))
+    ... def fn(seglst, *, speaker='X'):
+    ...     return seglst.map(lambda x: {**x, 'speaker': speaker})
+    >>> from meeteval.io.stm import STM
+    >>> fn(STM.parse('X 1 A 0 1 a b c'))
+    STM(lines=[STMLine(filename='X', channel=1, speaker_id='X', begin_time=0, end_time=1, transcript='a b c')])
+    >>> from meeteval.io.rttm import RTTM
+    >>> fn(RTTM.parse('SPEAKER CMU_20020319-1400_d01_NONE 1 130.430000 2.350 <NA> <NA> juliet <NA> <NA>'))
+    RTTM(lines=[RTTMLine(type='SPEAKER', filename='CMU_20020319-1400_d01_NONE', channel='1', begin_time=130.43, duration=2.3499999999999943, othography='<NA>', speaker_type='<NA>', speaker_id='X', confidence='<NA>', signal_look_ahead_time='<NA>')])
```

Comment:
We should switch to decimal as default representation for numbers to avoid:
 - Loss of visual precision: `130.430000` vs `130.43`
 - Issue that float has base 2: `2.350` vs `2.3499999999999943`

Reply:
Done